src/Xrefcheck/Progress.hs

```diff
@@ -22,59 +28,127 @@ import Universum
   } deriving stock (Show)
 
 -- | Initialise null progress.
 initProgress :: Num a => a -> Progress a
-initProgress a = Progress{ pTotal = a, pCurrent = 0, pErrors = 0 }
+initProgress a = Progress{ pTotal = a
+                         , pCurrent = 0
+                         , pErrorsUnfixable = 0
+                         , pErrorsFixable = 0
+                         , pTaskTimestamp = Nothing
+                         }
 
 -- | Increase progress amount.
 incProgress :: (Num a, Show a) => Progress a -> Progress a
 incProgress Progress{..} = Progress{ pCurrent = pCurrent + 1, .. }
 
--- | Increase errors amount.
-incProgressErrors :: (Num a, Show a) => Progress a -> Progress a
-incProgressErrors Progress{..} = Progress{ pErrors = pErrors + 1, .. }
+-- | Increase the number of unfixable errors.
+incProgressUnfixableErrors :: (Num a, Show a) => Progress a -> Progress a
+incProgressUnfixableErrors Progress{..} = Progress{ pErrorsUnfixable = pErrorsUnfixable + 1
+                                                  , ..
+                                                  }
+
+-- | Increase the number of fixable errors.
+incProgressFixableErrors :: (Num a, Show a) => Progress a -> Progress a
+incProgressFixableErrors Progress{..} = Progress{ pErrorsFixable = pErrorsFixable + 1
+                                                , ..
+                                                }
+
+-- | Decrease the number of fixable errors. This function indicates the situation where one of
+-- such errors had been successfully eliminated.
+decProgressFixableErrors :: (Num a, Show a) => Progress a -> Progress a
+decProgressFixableErrors Progress{..} = Progress{ pErrorsFixable = pErrorsFixable - 1
+                                                , ..
+                                                }
+
+fixableToUnfixable :: (Num a, Show a) => Progress a -> Progress a
+fixableToUnfixable Progress{..} = Progress{ pErrorsFixable = pErrorsFixable - 1
+                                          , pErrorsUnfixable = pErrorsUnfixable + 1
+                                          , ..
+                                          }
+
+setTaskTimestamp :: Time Second -> Time Second -> Progress a -> Progress a
+setTaskTimestamp ttc startTime Progress{..} = Progress{ pTaskTimestamp = Just (ttc, startTime)
+                                                      , ..
+                                                      }
+
+removeTaskTimestamp :: Progress a -> Progress a
+removeTaskTimestamp Progress{..} = Progress{ pTaskTimestamp = Nothing
+                                           , ..
+                                           }
+
+checkTaskTimestamp :: Time Second -> Progress a -> Progress a
+checkTaskTimestamp posixTime p@Progress{..} =
+  let mWithinGivenTime = do
+        (ttc, startTime) <- pTaskTimestamp
+        return $ ttc >= posixTime -:- startTime
+  in maybe p (\withinGivenTime ->
+    if withinGivenTime
+    then p
+    else removeTaskTimestamp p) mWithinGivenTime
```

Comment:
Ah, yeah, didn't notice the redundancy of `mWithinGivenTime` & its `maybe`-elimination afterward.

What about the following workaround?
```
maybe p (\(ttc, startTime) -> 
  if ttc >= posixTime -:- startTime
  then p
  else removeTaskTimestamp p) pTaskTimestamp
```
It employs `maybe`-elimination as well, instead of the explicit `Maybe`-pattern-matching.

Reply:
Personally, I prefer the explicit pattern matching. I think `maybe` is ok when the expression is small enough (e.g. `maybe 0 (+1) myvar`). But otherwise, the parenthesis and the fact that `pTaskTimestamp` is the very last argument just seem to get in the way, IMO.

But that's all subjective, it's just my personal opinion, so it's up to you really whichever you prefer ^^

Reply:
I guess if I tried to distil my thoughts, it mostly comes down to this: `case mx of Just x -> f 1; Nothing -> y` can be comfortably read from left-to-right, whereas `maybe y f mx` has to be read in an awkward order (I first look to the very right to see `mx`, then left to see `y`, then the middle to see `f`). When the `f` expression is big enough, this starts getting harder to read.

Reply:
I think I get it. I mean, indeed in this situation explicit pattern-matching on the `Maybe` construct is the most readable solution, and I reckon I have once again chased after the number of rows instead of the readability. 
Pattern-matching it is.